numpy/core/src/umath/ufunc_object.c

```diff
@@ -3239,17 +3239,16 @@ PyUFunc_Accumulate(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
         NPY_BEGIN_THREADS_NDITER(iter);
 
         do {
-
             dataptr_copy[0] = dataptr[0];
             dataptr_copy[1] = dataptr[1];
             dataptr_copy[2] = dataptr[0];
 
             /* Copy the first element to start the reduction */
             if (otype == NPY_OBJECT) {
+                Py_XINCREF(*(PyObject **)dataptr_copy[1]);
                 Py_XDECREF(*(PyObject **)dataptr_copy[0]);
```

Comment:
So to check if I understand: what's going on here is, we're doing an accumulate, `dataptr_copy[0]` points to the `out` array, `dataptr_copy[1]` points to the `in` array, and when `out is in` then bad stuff can happen. In that case I think this logic is now correct, though it is very tricky :-).

Observation 1: this desperately needs a comment :-).

Observation 2: could we instead replace this whole `if/else` stuff with a call to the `dtype`'s data copy method? that would let us remove the special case, we'll eventually need it when dtypes get smarter, and object dtype method should already handle this case?